### PR TITLE
Dev ppo

### PR DESCRIPTION
**Pull request by lhnguyen102**
Features
- Handling the continuous actions
- Single env with a single agent
- Multi-env with a single agent
- mujoco env

Run command for mujoco
```python -m main +experiment=ppo/hopper```

### CODE REVIEW

**Comment by saikrishna-1996** [comment]
Do we need the file `profile_stats_0.stats` ?

**Comment by lhnguyen102** [comment]
@saikrishna-1996 You're right. We don't need this file `profile_stats_0.stats`. I will remove it. Thank you!

File: .editorconfig [hunk 1]
```diff
@@ -7,7 +7,8 @@ charset = utf-8
 end_of_line = lf
 trim_trailing_whitespace = true
 insert_final_newline = true
-max_line_length = 120
+max_line_length = 80
+# max_line_length = 120
```
Comment (vabdollahi) [inline .editorconfig:11]: Please remove the commented line

File: cogment_verse/run/run_session.py [hunk 2]
```diff
@@ -24,7 +26,7 @@
 
 
 class RunSession:
-    def __init__(self, run_cfg, run_id, services_directory, model_registry):
+    def __init__(self, run_cfg, run_id, services_directory, model_registry: ModelRegistry):
```
Comment (vabdollahi) [inline cogment_verse/run/run_session.py:29]: Even though I like the idea of adding object types, in the cogment-verse it looks like an inconsistent coding style.

File: config/experiment/simple_a2c/cartpole.yaml [hunk 3]
```diff
@@ -2,15 +2,15 @@
...
   - override /services/environment: cartpole
 
 run:
   class_name: actors.simple_a2c.SimpleA2CTraining
   seed: 618
   num_epochs: 500
-  epoch_num_trials: 10
-  num_parallel_trials: 8
+  epoch_num_trials: 1
+  num_parallel_trials: 1
```
Comment (vabdollahi) [inline config/experiment/simple_a2c/cartpole.yaml]: If it does run in parallel, please use a larger value, if not, please remove it.

File: cogment_verse/constants.py [hunk 4]
```diff
@@ -12,7 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-COGMENT_VERSION = "v2.5.0"
+COGMENT_VERSION = "v2.8.0"
```
Comment (vabdollahi) [inline cogment_verse/constants.py:15]: I think 2.9 is out :)

File: actors/ppo.py [hunk 5]
```diff
@@ -0,0 +1,720 @@
...
+    AgentConfig,
+    EnvironmentConfig,
+    EnvironmentSpecs,
+    PlayerAction,
+    cog_settings,
+    flatten,
+    flattened_dimensions,
+    unflatten,
+)
+
```
Comment (vabdollahi) [inline actors/ppo.py:38]: We might need "torch.multiprocessing.set_sharing_strategy("file_system")" for Ubuntu

File: actors/ppo.py [hunk 6]
```diff
@@ -0,0 +1,720 @@
...
+        return new_mean, new_var, new_count
+
+
+class PPOModel(Model):
+    """Proximal Policy Optimization (PPO) is an on-policy algorithm.
+    https://arxiv.org/pdf/1707.06347.pdf.
+
+    Attributes:
+        model_id: Model identity
+        environment_implementation: Environment specs (not sure what it is)
```
Comment (vabdollahi) [inline actors/ppo.py:164]: not sure what it is? :D

File: actors/ppo.py [hunk 7]
```diff
@@ -0,0 +1,720 @@
...
+
+                    # Reset the data storage
+                    observations = []
+                    actions = []
+                    rewards = []
+                    dones = []
+                    if iter_idx % 1 == 0:
+                        # Compute average rewards for last 100 episodes
+                        avg_rewards = await self.compute_average_reward(episode_rewards)
+                        log.warning(
```
Comment (vabdollahi) [inline actors/ppo.py:522]: log.info?

File: actors/ppo.py [hunk 8]
```diff
@@ -0,0 +1,720 @@
...
+        num_epochs: int,
+    ) -> Tuple[torch.Tensor]:
+        """Update policy & value networks"""
+
+        returns = advs + values
+        num_states = len(returns)
+        for _ in range(num_epochs):
+            for _ in range(num_states // self._cfg.batch_size):
+                # Get data in batch. TODO: Send data to device (need to test with cuda)
+                idx = np.random.randint(0, num_states, self._cfg.batch_size)
```
Comment (vabdollahi) [inline actors/ppo.py:602]: You may create a rng = np.random.default_rng() outside the loop and use it here

File: actors/ppo.py [hunk 9]
```diff
@@ -0,0 +1,720 @@
...
+                state = states[idx].to(self._device)
+                action = actions[idx].to(self._device)
+                return_ = returns[idx].to(self._device)
+                adv = advs[idx].to(self._device)
+                # old_value = values[idx].to(self._device)
```
Comment (vabdollahi) [inline actors/ppo.py:607]: Remove the comment

**Comment by vabdollahi** [comment]
Being a non-expert in PPO, the implementation seems neat to me.

File: config/experiment/simple_a2c/cartpole.yaml [hunk 10]
```diff
@@ -2,5 +2,5 @@
 defaults:
   - override /services/actor:
       - simple_a2c
-  - override /run: simple_a2c
+  - override /run: headless_play
```
Comment (saikrishna-1996) [inline config/experiment/simple_a2c/cartpole.yaml:5]: Why?

File: package-lock.json [hunk 11]
```diff
@@ -0,0 +1,2382 @@
+{
```
Comment (saikrishna-1996) [inline package-lock.json:1]: Is this file needed?

Comment (saikrishna-1996) [reply package-lock.json:1]: same question for package.json

File: actors/ppo.py [hunk 12]
```diff
@@ -0,0 +1,720 @@
...
+        x = torch.tanh(x)
+
+        # Hidden layer
+        x = self.fully_connected(x)
+        x = torch.tanh(x)
+
+        # Output layer
+        mean = self.mean(x)
+        std = self.log_std.exp()
+        dist = torch.distributions.normal.Normal(mean, std)
```
Comment (saikrishna-1996) [inline actors/ppo.py:65]: shouldn't this be categorical (for discrete actions) and DiagGaussian (for continuous) ?

File: actors/ppo.py [hunk 13]
```diff
@@ -0,0 +1,720 @@
...
+log = logging.getLogger(__name__)
+
+# pylint: disable=E1102
+# pylint: disable=W0212
+class PolicyNetwork(torch.nn.Module):
+    """Gaussian policy network"""
+
+    def __init__(self, num_input: int, num_output: int, num_hidden: int) -> None:
+        super().__init__()
+        self.input = torch.nn.Linear(num_input, num_hidden)
```
Comment (saikrishna-1996) [inline actors/ppo.py:48]: a recent implementation also recommends layer initialization: https://github.com/vwxyzjn/ppo-implementation-details/blob/main/ppo.py#L95

File: actors/ppo.py [hunk 14]
```diff
@@ -0,0 +1,720 @@
...
+        # Learning schedule
+        self.policy_scheduler = torch.optim.lr_scheduler.ExponentialLR(self.policy_optimizer, gamma=0.99)
+        self.value_scheduler = torch.optim.lr_scheduler.ExponentialLR(self.value_optimizer, gamma=0.99)
+
+        # version user data
+        self.iter_idx = 0
+        self.total_samples = 0
+
+    @property
+    def state_norm(self) -> bool:
```
Comment (saikrishna-1996) [inline actors/ppo.py:246]: Do we need two `state_norm` functions ?

File: actors/ppo.py [hunk 15]
```diff
@@ -0,0 +1,720 @@
...
+        """Get actor"""
+        return [PLAYER_ACTOR_CLASS]
+
+    async def impl(self, actor_session):
+        # Start a session
+        actor_session.start()
+        config = actor_session.config
+        assert config.environment_specs.num_players == 1
+        assert len(config.environment_specs.action_space.properties) == 1
+        assert config.environment_specs.action_space.properties[0].WhichOneof("type") == "box"
```
Comment (saikrishna-1996) [inline actors/ppo.py:312]: Ideally PPO should also work for discrete and multi-discrete actions

File: actors/ppo.py [hunk 16]
```diff
@@ -0,0 +1,720 @@
...
+        "grad_norm": 0.5,
+        "state_norm": False,
+    }
+
+    def __init__(self, environment_specs: EnvironmentSpecs, cfg: EnvironmentConfig) -> None:
+        super().__init__()
+        self._dtype = torch.float
+        self._environment_specs = environment_specs
+        self._cfg = cfg
+        self.mse_loss = torch.nn.MSELoss()
```
Comment (saikrishna-1996) [inline actors/ppo.py:387]: this wasn't used anywhere?

File: environments/gym_wrapper.py [hunk 17]
```diff
@@ -0,0 +1,118 @@
+# Copyright 2022 AI Redefined Inc. <dev+cogment@ai-r.com>
```
Comment (saikrishna-1996) [inline environments/gym_wrapper.py:1]: Could you explain why we need this or the differences with gym_adapter?

File: .editorconfig (hunk 1 above)
Comment (lhnguyen102) [reply .editorconfig:11]: Done

File: cogment_verse/run/run_session.py (hunk 2 above)
Comment (lhnguyen102) [reply cogment_verse/run/run_session.py:29]: I agree it won't be consistent for now but I think we should move towards the hint type in order to facilitate the code navigation using one click instead of searching option, especially for new users.

File: config/experiment/simple_a2c/cartpole.yaml (hunk 3 above)
Comment (lhnguyen102) [reply config/experiment/simple_a2c/cartpole.yaml]: I've changed to larger values but this choice should be selected by the users on whether if they want to run 
- one trial -> one agent
- multiple trial -> one agent 

The reason is because PPO is on-policy RL algorithm. It is a bit different from DQN i.e., off-policy RL algorithm

File: cogment_verse/constants.py (hunk 4 above)
Comment (lhnguyen102) [reply cogment_verse/constants.py]: For some reasons, 2.9.0 has not worked yet :)

File: actors/ppo.py (hunk 5 above)
Comment (lhnguyen102) [reply actors/ppo.py:38]: I added this command line to `ppo.py`. Please test it out if it works

File: actors/ppo.py (hunk 6 above)
Comment (lhnguyen102) [reply actors/ppo.py:164]: I have replaced with more descriptive comments :)

File: actors/ppo.py (hunk 7 above)
Comment (lhnguyen102) [reply actors/ppo.py:522]: yes and done!

File: actors/ppo.py (hunk 8 above)
Comment (lhnguyen102) [reply actors/ppo.py:602]: I think ` np.random.default_rng()` for a wide range of distributions while here I only target the discrete uniform for the interger only

File: actors/ppo.py (hunk 9 above)
Comment (lhnguyen102) [reply actors/ppo.py:607]: Done

File: config/experiment/simple_a2c/cartpole.yaml (hunk 10 above)
Comment (lhnguyen102) [reply config/experiment/simple_a2c/cartpole.yaml:5]: It is just leftover from the previous version of the next branch. I have changed ` override /run: headless_play` to `override /run: simple_a2c`

File: package-lock.json (hunk 11 above)
Comment (lhnguyen102) [reply package-lock.json:1]: No, you're right. It has been removed

File: actors/ppo.py (hunk 12 above)
Comment (lhnguyen102) [reply actors/ppo.py:65]: In this version, I only focus on the continuous actions. I will implement discrete actions, a mix of discrete and continuous, and multi discrete in the future version

File: actors/ppo.py (hunk 13 above)
Comment (lhnguyen102) [reply actors/ppo.py:48]: Yes, it has been implemented in `ppo.py` (line 225-226) where the orthogonal initialization has been used

File: actors/ppo.py (hunk 15 above)
Comment (lhnguyen102) [reply actors/ppo.py:312]: Agree. Upcoming version will address this issue

File: actors/ppo.py (hunk 16 above)
Comment (lhnguyen102) [reply actors/ppo.py:387]: You're right. It has been removed

File: environments/gym_wrapper.py (hunk 17 above)
Comment (lhnguyen102) [reply environments/gym_wrapper.py:1]: It is because in the `gym_adapter.py` leans towards the dAgger algorithm where it has TEACHER actor et RL actor. @cloderic asked me to implement a vanilla version for gym envs (I basically remove TEACHER  actor).
